src/users/users.service.spec.ts

```diff
@@ -31,15 +34,124 @@ describe('UsersService', () => {
         }),
         LoggerModule,
       ],
-    })
-      .overrideProvider(getRepositoryToken(User))
-      .useValue({})
-      .compile();
+    }).compile();
 
     service = module.get<UsersService>(UsersService);
+    userRepo = module.get<Repository<User>>(getRepositoryToken(User));
   });
 
   it('should be defined', () => {
     expect(service).toBeDefined();
   });
+
+  describe('createUser', () => {
+    const username = 'hardcoded';
+    const displayname = 'Testy';
+    beforeEach(() => {
+      jest
+        .spyOn(userRepo, 'save')
+        .mockImplementationOnce(async (user: User): Promise<User> => user);
+    });
+    it('works', async () => {
+      const user = await service.createUser(username, displayname);
+      expect(user.userName).toEqual(username);
+      expect(user.displayName).toEqual(displayname);
+    });
+    it('fails if username is already taken', async () => {
+      jest.spyOn(userRepo, 'save').mockImplementationOnce(() => {
+        throw new Error();
+      });
+      // create first user with username
+      await service.createUser(username, displayname);
+      // attempt to create second user with username
+      await expect(service.createUser(username, displayname)).rejects.toThrow(
+        AlreadyInDBError,
+      );
+    });
+  });
+
+  describe('deleteUser', () => {
+    it('works', async () => {
+      const username = 'hardcoded';
+      const displayname = 'Testy';
+      const user = User.create(username, displayname) as User;
+      jest.spyOn(userRepo, 'remove').mockImplementationOnce(
+        // eslint-disable-next-line @typescript-eslint/require-await
+        async (user: User): Promise<User> => {
+          expect(user).toEqual(user);
```

Comment:
I think this does not what it is supposed to do. The second `user` variable covers the other one.

Reply:
good point. Changed…